File: integrations/claude-managed-agents/python/ag_ui_claude_managed_agents/sessions.py
"""Thread-to-session stores."""

from collections import OrderedDict
from copy import deepcopy

from .constants import IN_MEMORY_SESSION_STORE_MAX_ENTRIES
from .types import SessionRecord


class InMemorySessionStore:
    """In-memory thread-to-session store. Mappings are lost on restart.

    Bounded: thread ids come from the client, so an unbounded dict is a memory
    leak an untrusted caller controls. Once `max_entries` is reached the
    least-recently-used mapping is dropped -- which costs that thread its
    history (the next run starts a fresh session), so a deployment that cannot
    afford that should supply a persistent store instead.
    """

    def __init__(self, max_entries: int = IN_MEMORY_SESSION_STORE_MAX_ENTRIES) -> None:
        if max_entries < 1:
            raise ValueError("max_entries must be a positive integer")
        self._max_entries = max_entries
        self._records: OrderedDict[str, SessionRecord] = OrderedDict()

    def get(self, key: str) -> SessionRecord | None:
        record = self._records.get(key)
        if record is None:
            return None
        self._records.move_to_end(key)
        # Hand out a copy: the agent mutates records in place between
        # persists, so an aliased record would make an unpersisted mutation
        # indistinguishable from a persisted one — and a dropped write would
        # only surface against a real out-of-process store.
        return deepcopy(record)

    def set(self, key: str, record: SessionRecord) -> None:
        self._records[key] = deepcopy(record)
        self._records.move_to_end(key)
        while len(self._records) > self._max_entries:
            self._records.popitem(last=False)

    def delete(self, key: str) -> None:
        self._records.pop(key, None)

    def __len__(self) -> int:
        """How many mappings are currently held."""
        return len(self._records)

File: integrations/claude-managed-agents/typescript/src/sessions.ts
import { IN_MEMORY_SESSION_STORE_MAX_ENTRIES } from "./constants";
import type { SessionRecord, SessionStore } from "./types";

/**
 * A defensive copy of a record.
 *
 * The store must not hand out a reference to the record it holds: the agent
 * mutates records in place between persists, so an aliased record would make
 * an unpersisted mutation indistinguishable from a persisted one — and a
 * dropped write would only surface against a real out-of-process store.
 */
const copy = (record: SessionRecord): SessionRecord => structuredClone(record);

/**
 * In-memory thread↔session store. Mappings are lost on restart.
 *
 * Bounded: thread ids come from the client, so an unbounded map is a memory
 * leak an untrusted caller controls. Once `maxEntries` is reached the
 * least-recently-used mapping is dropped — which costs that thread its history
 * (the next run starts a fresh session), so a deployment that cannot afford
 * that should supply a persistent store instead.
 */
export class InMemorySessionStore implements SessionStore {
  // Insertion order is the recency order: a read or write re-inserts its key,
  // so the oldest entry is always the first one Map iteration yields.
  private records = new Map<string, SessionRecord>();

  constructor(private readonly maxEntries: number = IN_MEMORY_SESSION_STORE_MAX_ENTRIES) {
    if (!Number.isInteger(maxEntries) || maxEntries < 1) {
      throw new RangeError("maxEntries must be a positive integer");
    }
  }

  get(key: string): SessionRecord | undefined {
    const record = this.records.get(key);
    if (record === undefined) return undefined;
    this.records.delete(key);
    this.records.set(key, record);
    return copy(record);
  }

  set(key: string, record: SessionRecord): void {
    this.records.delete(key);
    this.records.set(key, copy(record));
    while (this.records.size > this.maxEntries) {
      const oldest = this.records.keys().next();
      if (oldest.done) break;
      this.records.delete(oldest.value);
    }
  }

  delete(key: string): void {
    this.records.delete(key);
  }

  /** How many mappings are currently held. */
  get size(): number {
    return this.records.size;
  }
}
